### DISCUSSION TITLE
Force Flyout/Popup Load in ControlTemplates

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#9128] [General]
Twice now there have been issues with the ColorPicker I can't solve without some ugly work-arounds.

 * #9098
 * https://github.com/AvaloniaUI/Avalonia/pull/8919#discussion_r967715869

Both of these are caused by the Popup in the control template not being loaded until just before it opens. I don't believe WPF or WinUI work this way -- or I was just lucky and never ran into an issue there. I would like the Popup to load normally as with all elements in the ControlTemplate. My question is how to do this:

 1. Is there a way to force or trick Avalonia into loading the Popup?
 2. Can this be changed in the general case to always load all control template elements including popups? A control is a self-contained part of the code that really shouldn't need to worry about this.

**Comment by robloo** [comment]
@amwx This might actually be the Flyout functionality instead of the Popup. Would you happen to know if there is an easy way to adjust this?

**Reply by amwx** [reply]
So, I don't think these issues are related to the popup/flyout loading late. 

For the first issue:
I took the flyout out of the `ColorPicker` template and tried setting the color and it didn't take. The following still resulted in the `ColorPicker` showing white as the current color.

```Xaml
<ColorPicker Color="PaleVioletRed" />
```

This is caused by the `ignorePropertyChanged` check in `OnPropertyChanged` of the `ColorView`. 
https://github.com/AvaloniaUI/Avalonia/blob/c24b0090e76cfae9973a3b866c7f4bf78828660f/src/Avalonia.Controls.ColorPicker/ColorView/ColorView.cs#L201-L222

The initial setting of the `Color` property is ignored because `ignorePropertyChanged == true` when the ColorPicker gets that notification so it's ignored and isn't set. I commented that check out and the color properly set. Which also appeared in the flyout when I returned that to the template. So this just a needs a little refactoring to be smarter about when to actually ignore the property change.

Before/After commenting that out:
![image](https://user-images.githubusercontent.com/40413319/194723413-1ffd534f-363c-40fe-a544-906670d260c3.png)

**Reply by amwx** [reply]
(splitting because of length)
As for the second issue:
This one is interesting. I'm still not 100% sure what's going on here, but here's what I've figure out so far:
Setting `Palette` to the `ColorPicker` does work (received in OnPropertyChanged)
When the ColorPicker is opened, there are a following 4 OnPropertyChanged notifications for `PaletteProperty`:

```
1. Setting Palette property with Style priority. NewValue: Avalonia.Controls.FluentColorPalette, TemplatedParent: null
2. Setting Palette property with Unset priority. NewValue: , TemplatedParent: null
3. Setting Palette property with Style priority. NewValue: Avalonia.Controls.FluentColorPalette, TemplatedParent: null
4. Setting Palette property with TemplatedParent priority. NewValue: Avalonia.Controls.FlatColorPalette, TemplatedParent: Avalonia.Controls.ColorPicker
```
The first one is when the template is loaded - it correctly detects that.
The second one is interesting - notice the `Unset` priority. It turns out that this is a result of the popup being detached from the logical tree. Walking up the call stack this happens when ` ((ISetLogicalParent)popupHost).SetParent(this);` is set in `Popup` - apparently this is triggering a detachment event (more on this in a second). This detaches all styles. This is a bug, but is probably unrelated.
The third one is when the popup is reattached from having its logical parent set, so all styles are reapplied and this is equivalent to step 1, we're reading the value set in the template (again).
For the fourth one, `TemplatedParent` is finally set and it reads the value from `ColorPicker`. Except, this value doesn't actually seem to take because it shows the correct value and priority, which is higher than `Style`, except: the Palette view still shows the `FluentColorPalette`.

If you inspect up to `EndBatchUpdate` after (2) above, there is one value in the property store for `Palette`, which is {empty} with TemplatedParent priority - which is correct since TemplatedParent hasn't been set. After (3), there are two values in the property store, value [0] is the Style priority from the template and value [1] is the {empty} templated parent property.
After (4), there's no batch update here, but you can see the property store in `PriorityValue.UpdateEffectiveValue` as part of the binding evaluation. Interestingly, there's only 1 entry here, the value set from reading the style. It doesn't appear that this TemplatedParent binding was ever written to the property store (walking the call stack, I don't see where its written, only where the property change notification is assembled) - so when the control finishes loading, it only sees the `FluentColorPalette` from the style - but never the value it was supposed to inherit from `ColorPicker`. I'm not all that familiar with the property system backend but something is going way wrong here. So I don't know if there's a bug here or if I'm missing something.

**Reply by timunie** [reply]
Let me add two things:

1. ValueStore is going to be changed. There is a PR for this if not ywt merged. Maybe one should double check if this PR solves the issue. 

2. When it comes to WPF I also had an issue where the popup was not yet loaded which breaks some functionality until today. If you are interested, I can link the control I'm talking about here. Just let me know.

**Reply by robloo** [reply]
@amwx Thanks for looking into this in such detail. It's definitely unexpected and much more detailed than any debugging I did so far... I will step through the code with your comments above and see if I can make more sense of it.

1. If the Color property is not working in just a ColorView.... well I should be able to fix that and also should have seen it earlier.
2. Your understanding of the styling system internals is much more advanced than mine so not sure how far I'll get on the second one.

**Reply by robloo** [reply]
I stepped through the code again using @amwx idea of removing the flyout entirely. However, it confirmed my earlier tests. It isn't a simple refactoring fix or really much of an issue with `ignorePropertyChanges`. The issue is simply lifecycle:

The ColorPicker is also a ColorView. However, there is also a ColorView in the control template. The control template's ColorView was intended to be the main one and do the heavy lifting. Therefore, the ColorPicker just disabled property change handling and let's the ColorView in the template do all the work. That works find -- as long as the template has been applied and the ColorView exists. However, when the ColorPicker control is first created any properties specified in XAML or code-behind (like Color) are set BEFORE the template is applied. This then causes the problem here because someone has to do the processing in this case but no one is: The ColorPicker ignored the changes expecting it's bound value to be passed to the template's ColorView and handled. Meanwhile the template ColorView doesn't even exist yet.

Anyway, at least for issue 1 above it's confirmed the flyout is not the problem. So i'll consider that part of this discussion closed. The second part about the palette not applying correctly may still be an issue here but I'll wait and retest with the new value store before filing a separate issue.

In the meantime I'm going to have to 1. Make the ColorPicker the primary control instead of deferring to the template's ColorView or 2. Be smarter about handing off control once the template is applied.
